src/poly/commitment.rs

```diff
@@ -154,6 +217,95 @@ impl<C: CurveAffine> Params<C> {
     }
 }
 
+/// A guard returned by the verifier
+#[derive(Debug)]
+pub struct Guard<'a, C: CurveAffine> {
+    /// Vector of random generators
+    pub g: Vec<C>,
+
+    /// Random generator
+    pub h: C,
+
+    /// Negation of z1 value in the OpeningProof
+    pub neg_z1: C::Scalar,
+
+    /// Params that were used by the verifier
+    pub params: &'a Params<C>,
+
+    /// Scalars produced by the verifier for multiscalar multiplication
+    pub scalars: Vec<C::Scalar>,
+
+    /// Points produced by the verifier for multiscalar multiplication
+    pub bases: Vec<C>,
+}
+
+impl<'a, C: CurveAffine> Guard<'a, C> {
+    /// Lets caller supply the challenges and obtain an MSM with updated
+    /// scalars and points.
+    pub fn use_challenges(
+        &mut self,
+        challenges_sq_packed: Vec<Challenge>,
+    ) -> Result<MSM<C>, Error> {
+        // - [z1] G
+        let mut allinv = C::Scalar::one();
+        let mut challenges_sq = Vec::with_capacity(self.params.k as usize);
+
+        for challenge_sq_packed in challenges_sq_packed {
+            let challenge_sq: C::Scalar = get_challenge_scalar(challenge_sq_packed);
+            challenges_sq.push(challenge_sq);
+
+            let challenge = challenge_sq.deterministic_sqrt();
+            if challenge.is_none() {
+                // We didn't sample a square.
+                return Err(Error::OpeningError);
+            }
+            let challenge = challenge.unwrap();
+
+            let challenge_inv = challenge.invert();
```

Comment:
Instead of inverting here, you could have a `challenges` vector outside of the loop that you push to. Then, you could do `let allinv = C::Scalar::batch_invert(&mut challenges);` to batch the inversions together.

Reply:
Alternatively, we don't need to compute this again. It can all be stored in the guard object.

Reply:
i'm storing `allinv` in `Guard` to be used in `use_g()`. but i don't think it should be used in `use_challenges` because `allinv` is derived from `challenges` which should be provided by the caller here.